src/toxcore/dht/server.rs

```diff
@@ -170,26 +188,33 @@ impl Server {
         match packet {
             DhtPacket::PingRequest(packet) => {
                 debug!("Received ping request");
-                let client = self.create_client(&addr, packet.pk);
+                let state = self.state.read();
+                let client = self.create_client(&addr, packet.pk, &state.peers_cache);
                 self.handle_ping_req(client, packet)
             },
             DhtPacket::PingResponse(packet) => {
                 debug!("Received ping response");
-                if let Some(client) = self.get_client(&packet.pk) {
-                    self.handle_ping_resp(client, packet)
+                let mut state = self.state.write();
+                if let Some(mut client) = self.get_client(&packet.pk, &state.peers_cache) {
+                    let result = self.handle_ping_resp(client.clone(), packet.clone(), &mut state.kbucket);
+                    client.last_resp_time = Instant::now();
+                    state.peers_cache.insert(packet.pk, client);
+                    result
                 } else { // If there doesn't exist client in hash, then the PingRequest sent is not from me. Do nothing.
                     Box::new(future::ok(()))
                 }
             },
             DhtPacket::NodesRequest(packet) => {
                 debug!("Received NodesRequest");
-                let client = self.create_client(&addr, packet.pk);
-                self.handle_nodes_req(client, packet)
+                let state = self.state.read();
+                let client = self.create_client(&addr, packet.pk, &state.peers_cache);
+                self.handle_nodes_req(client, packet, &state.kbucket)
```

Comment:
Why don't you
1. call self.handle_nodes_req(packet)
2. inside handle_nodes_req you 1) take read lock 2) create client. 3) continue processing?